pkg/sql/catalog/tabledesc/constraint.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tabledesc
+
+import (
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
+)
+
+// constraintCache contains precomputed slices of constraints, categorized by kind and validity.
+// A constraint is considered
+// - active if its validity is VALIDATED;
+// - inactive if its validity is VALIDATING or UNVALIDATED;
+// - dropping if its validity is DROPPING;
+type constraintCache struct {
+	all                  []catalog.Constraint
+	allActiveAndInactive []catalog.Constraint
+	allActive            []catalog.Constraint
+
+	allChecks                  []catalog.Constraint
+	allActiveAndInactiveChecks []catalog.Constraint
+	allActiveChecks            []catalog.Constraint
+
+	allNotNulls                  []catalog.Constraint
+	allActiveAndInactiveNotNulls []catalog.Constraint
+	allActiveNotNulls            []catalog.Constraint
+
+	allFKs                  []catalog.Constraint
+	allActiveAndInactiveFKs []catalog.Constraint
+	allActiveFKs            []catalog.Constraint
+
+	allUniqueWithoutIndexs                  []catalog.Constraint
+	allActiveAndInactiveUniqueWithoutIndexs []catalog.Constraint
+	allActiveUniqueWithoutIndexs            []catalog.Constraint
+}
+
+// newConstraintCache returns a fresh fully-populated constraintCache struct for the
+// TableDescriptor.
+func newConstraintCache(desc *descpb.TableDescriptor, mutations *mutationCache) *constraintCache {
+	c := constraintCache{}
+
+	// addIfNotExists is a function that adds constraint `c` to slice `dest` if this
+	// constraint is not already in it (as determined by using a set of already added
+	// constraint IDs `constraintIDsInDest`).
+	// If `constraintIDsInDest` is nil, blindly append `c` to `dest`.
+	addIfNotExists := func(
+		c catalog.Constraint,
+		dest []catalog.Constraint,
+		constraintIDsInDest map[descpb.ConstraintID]bool,
+	) []catalog.Constraint {
+		if constraintIDsInDest == nil {
+			dest = append(dest, c)
+			return dest
+		}
+
+		if _, exist := constraintIDsInDest[c.GetConstraintID()]; exist {
+			return dest
+		}
+		dest = append(dest, c)
+		constraintIDsInDest[c.GetConstraintID()] = true
+		return dest
+	}
```

Comment:
Is it at all possible for a constraint to present itself more than once?

Reply:
yes, I knew of one scenario: when dropping a check constraint, we would find that constraint in the `Checks` slice, mark its validity to `DROPPING` *without* removing it from the `Checks` slice. We would also enqueue a mutation for this check constraint with `DROP` direction. Here, this dropping check constraint will be found both in `Checks` slice and on `Mutations` slice.

Reply:
Thanks! I feared it would be something like this.